# Language Translation
In this project, you’re going to take a peek into the realm of neural network machine translation.  You’ll be training a sequence to sequence model on a dataset of English and French sentences that can translate new sentences from English to French.
## Get the Data
Since translating the whole language of English to French will take lots of time to train, we have provided you with a small portion of the English corpus.

In [1]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import helper
import problem_unittests as tests

source_path = 'data/small_vocab_en'
target_path = 'data/small_vocab_fr'
source_text = helper.load_data(source_path)
target_text = helper.load_data(target_path)

## Explore the Data
Play around with view_sentence_range to view different parts of the data.

In [2]:
view_sentence_range = (0, 10)

"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import numpy as np

print('Dataset Stats')
print('Roughly the number of unique words: {}'.format(len({word: None for word in source_text.split()})))

sentences = source_text.split('\n')
word_counts = [len(sentence.split()) for sentence in sentences]
print('Number of sentences: {}'.format(len(sentences)))
print('Average number of words in a sentence: {}'.format(np.average(word_counts)))

print()
print('English sentences {} to {}:'.format(*view_sentence_range))
print('\n'.join(source_text.split('\n')[view_sentence_range[0]:view_sentence_range[1]]))
print()
print('French sentences {} to {}:'.format(*view_sentence_range))
print('\n'.join(target_text.split('\n')[view_sentence_range[0]:view_sentence_range[1]]))

Dataset Stats
Roughly the number of unique words: 227
Number of sentences: 137861
Average number of words in a sentence: 13.225277634719028

English sentences 0 to 10:
new jersey is sometimes quiet during autumn , and it is snowy in april .
the united states is usually chilly during july , and it is usually freezing in november .
california is usually quiet during march , and it is usually hot in june .
the united states is sometimes mild during june , and it is cold in september .
your least liked fruit is the grape , but my least liked is the apple .
his favorite fruit is the orange , but my favorite is the grape .
paris is relaxing during december , but it is usually chilly in july .
new jersey is busy during spring , and it is never hot in march .
our least liked fruit is the lemon , but my least liked is the grape .
the united states is sometimes busy during january , and it is sometimes warm in november .

French sentences 0 to 10:
new jersey est parfois calme pendant l' automne 

## Implement Preprocessing Function
### Text to Word Ids
As you did with other RNNs, you must turn the text into a number so the computer can understand it. In the function `text_to_ids()`, you'll turn `source_text` and `target_text` from words to ids.  However, you need to add the `<EOS>` word id at the end of `target_text`.  This will help the neural network predict when the sentence should end.

You can get the `<EOS>` word id by doing:
```python
target_vocab_to_int['<EOS>']
```
You can get other word ids using `source_vocab_to_int` and `target_vocab_to_int`.

In [3]:
def text_to_ids(source_text, target_text, source_vocab_to_int, target_vocab_to_int):
    """
    Convert source and target text to proper word ids
    :param source_text: String that contains all the source text.
    :param target_text: String that contains all the target text.
    :param source_vocab_to_int: Dictionary to go from the source words to an id
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :return: A tuple of lists (source_id_text, target_id_text)
    """
    # TODO: Implement Function
#     idx = 0
#     for w in set([word for word in source_text.split(' ')]):
#         source_vocab_to_int[w] = idx
#         idx += 1
#     idx = 0
#     for w in set([word for word in target_text.split(' ')]):
#         target_vocab_to_int[w] = idx
#         idx += 1
#     return source_vocab_to_int, target_vocab_to_int
    source_id_text = []
    target_id_text = []
    print(source_text)
    for line in source_text.split('\n'):
        line_enc = []
        for word in line.split(' '):
            if word in source_vocab_to_int.keys():
                line_enc.append(source_vocab_to_int[word])
        source_id_text.append(line_enc)
        
    for line in target_text.split('\n'):
        line_enc = []
        for word in line.split(' '):
            if word in target_vocab_to_int.keys():
                line_enc.append(target_vocab_to_int[word])
        line_enc.append(target_vocab_to_int['<EOS>'])
        target_id_text.append(line_enc)
        
    
    return source_id_text, target_id_text

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_text_to_ids(text_to_ids)

new jersey is sometimes quiet during autumn , and it is snowy in april .
the united states is usually chilly during july , and it is usually freezing in november .
california is usually quiet during march , and it is usually hot in june .
the united states is sometimes mild during june , and it is cold in september .
Tests Passed


### Preprocess all the data and save it
Running the code cell below will preprocess all the data and save it to file.

In [4]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
helper.preprocess_and_save_data(source_path, target_path, text_to_ids)

IOPub data rate exceeded.
The notebook server will temporarily stop sending output
to the client in order to avoid crashing it.
To change this limit, set the config variable
`--NotebookApp.iopub_data_rate_limit`.

Current values:
NotebookApp.iopub_data_rate_limit=1000000.0 (bytes/sec)
NotebookApp.rate_limit_window=3.0 (secs)



# Check Point
This is your first checkpoint. If you ever decide to come back to this notebook or have to restart the notebook, you can start from here. The preprocessed data has been saved to disk.

In [5]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import numpy as np
import helper
import problem_unittests as tests

(source_int_text, target_int_text), (source_vocab_to_int, target_vocab_to_int), _ = helper.load_preprocess()

### Check the Version of TensorFlow and Access to GPU
This will check to make sure you have the correct version of TensorFlow and access to a GPU

In [6]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
from distutils.version import LooseVersion
import warnings
import tensorflow as tf
from tensorflow.python.layers.core import Dense

# Check TensorFlow Version
assert LooseVersion(tf.__version__) >= LooseVersion('1.1'), 'Please use TensorFlow version 1.1 or newer'
print('TensorFlow Version: {}'.format(tf.__version__))

# Check for a GPU
if not tf.test.gpu_device_name():
    warnings.warn('No GPU found. Please use a GPU to train your neural network.')
else:
    print('Default GPU Device: {}'.format(tf.test.gpu_device_name()))

TensorFlow Version: 1.3.0


/Users/shubhama/anaconda/lib/python3.5/site-packages/ipykernel/__main__.py:15: UserWarning: No GPU found. Please use a GPU to train your neural network.


## Build the Neural Network
You'll build the components necessary to build a Sequence-to-Sequence model by implementing the following functions below:
- `model_inputs`
- `process_decoder_input`
- `encoding_layer`
- `decoding_layer_train`
- `decoding_layer_infer`
- `decoding_layer`
- `seq2seq_model`

### Input
Implement the `model_inputs()` function to create TF Placeholders for the Neural Network. It should create the following placeholders:

- Input text placeholder named "input" using the TF Placeholder name parameter with rank 2.
- Targets placeholder with rank 2.
- Learning rate placeholder with rank 0.
- Keep probability placeholder named "keep_prob" using the TF Placeholder name parameter with rank 0.
- Target sequence length placeholder named "target_sequence_length" with rank 1
- Max target sequence length tensor named "max_target_len" getting its value from applying tf.reduce_max on the target_sequence_length placeholder. Rank 0.
- Source sequence length placeholder named "source_sequence_length" with rank 1

Return the placeholders in the following the tuple (input, targets, learning rate, keep probability, target sequence length, max target sequence length, source sequence length)

In [7]:
def model_inputs():
    """
    Create TF Placeholders for input, targets, learning rate, and lengths of source and target sequences.
    :return: Tuple (input, targets, learning rate, keep probability, target sequence length,
    max target sequence length, source sequence length)
    """
    # TODO: Implement Function
    input_ph = tf.placeholder(tf.int32, [None,None], name = 'input')
    targets = tf.placeholder(tf.int32, [None,None], name = 'targets')
    learning_rate = tf.placeholder(tf.float32, name = 'learning_rate')
    keep_prob = tf.placeholder(tf.float32, name = 'keep_prob')
    target_sequence_length = tf.placeholder(tf.int32, (None,), name = 'target_sequence_length')
    max_target_len = tf.reduce_max(target_sequence_length, name = 'max_target_len')
    source_sequence_length = tf.placeholder(tf.int32, (None,), name = 'source_sequence_length')
    return input_ph, targets, learning_rate, keep_prob, target_sequence_length, max_target_len, source_sequence_length


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_model_inputs(model_inputs)

ERROR:tensorflow:==================================
Object was never used (type <class 'tensorflow.python.framework.ops.Operation'>):
<tf.Operation 'assert_rank_2/Assert/Assert' type=Assert>
If you want to mark it as used call its "mark_used()" method.
It was originally created here:
['File "/Users/shubhama/anaconda/lib/python3.5/runpy.py", line 184, in _run_module_as_main\n    "__main__", mod_spec)', 'File "/Users/shubhama/anaconda/lib/python3.5/runpy.py", line 85, in _run_code\n    exec(code, run_globals)', 'File "/Users/shubhama/anaconda/lib/python3.5/site-packages/ipykernel/__main__.py", line 3, in <module>\n    app.launch_new_instance()', 'File "/Users/shubhama/anaconda/lib/python3.5/site-packages/traitlets/config/application.py", line 653, in launch_instance\n    app.start()', 'File "/Users/shubhama/anaconda/lib/python3.5/site-packages/ipykernel/kernelapp.py", line 477, in start\n    ioloop.IOLoop.instance().start()', 'File "/Users/shubhama/anaconda/lib/python3.5/site-packages/zm

Tests Passed


### Process Decoder Input
Implement `process_decoder_input` by removing the last word id from each batch in `target_data` and concat the GO ID to the begining of each batch.

In [8]:
def process_decoder_input(target_data, target_vocab_to_int, batch_size):
    """
    Preprocess target data for encoding
    :param target_data: Target Placehoder
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :param batch_size: Batch Size
    :return: Preprocessed target data
    """
    # TODO: Implement Function
    ending = tf.strided_slice(target_data, [0,0], [batch_size,-1],[1,1])
    output = tf.concat([tf.fill([batch_size, 1], target_vocab_to_int['<GO>']), ending], 1)
    return output

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_process_encoding_input(process_decoder_input)

Tests Passed


### Encoding
Implement `encoding_layer()` to create a Encoder RNN layer:
 * Embed the encoder input using [`tf.contrib.layers.embed_sequence`](https://www.tensorflow.org/api_docs/python/tf/contrib/layers/embed_sequence)
 * Construct a [stacked](https://github.com/tensorflow/tensorflow/blob/6947f65a374ebf29e74bb71e36fd82760056d82c/tensorflow/docs_src/tutorials/recurrent.md#stacking-multiple-lstms) [`tf.contrib.rnn.LSTMCell`](https://www.tensorflow.org/api_docs/python/tf/contrib/rnn/LSTMCell) wrapped in a [`tf.contrib.rnn.DropoutWrapper`](https://www.tensorflow.org/api_docs/python/tf/contrib/rnn/DropoutWrapper)
 * Pass cell and embedded input to [`tf.nn.dynamic_rnn()`](https://www.tensorflow.org/api_docs/python/tf/nn/dynamic_rnn)

In [9]:
from imp import reload
reload(tests)

def encoding_layer(rnn_inputs, rnn_size, num_layers, keep_prob, 
                   source_sequence_length, source_vocab_size, 
                   encoding_embedding_size):
    """
    Create encoding layer
    :param rnn_inputs: Inputs for the RNN
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param keep_prob: Dropout keep probability
    :param source_sequence_length: a list of the lengths of each sequence in the batch
    :param source_vocab_size: vocabulary size of source data
    :param encoding_embedding_size: embedding size of source data
    :return: tuple (RNN output, RNN state)
    """
    # TODO: Implement Function
    enc_embed_input = tf.contrib.layers.embed_sequence(rnn_inputs,source_vocab_size, 
                   encoding_embedding_size)
    # RNN cell
    def make_cell(rnn_size):
        enc_cell = tf.contrib.rnn.LSTMCell(rnn_size,initializer=tf.random_uniform_initializer(-0.1, 0.1, seed=2))
        drop_cell = tf.contrib.rnn.DropoutWrapper(enc_cell, output_keep_prob = keep_prob)
        return drop_cell

    # Construct a stacked tf.contrib.rnn.LSTMCell wrapped in a tf.contrib.rnn.DropoutWrapper
    enc_cell = tf.contrib.rnn.MultiRNNCell([make_cell(rnn_size) for _ in range(num_layers)])
    
    # Pass cell and embedded input to tf.nn.dynamic_rnn()
    enc_output, enc_state = tf.nn.dynamic_rnn(enc_cell, enc_embed_input, sequence_length=source_sequence_length, dtype=tf.float32)
    
    return enc_output, enc_state

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_encoding_layer(encoding_layer)

Tests Passed


### Decoding - Training
Create a training decoding layer:
* Create a [`tf.contrib.seq2seq.TrainingHelper`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/TrainingHelper) 
* Create a [`tf.contrib.seq2seq.BasicDecoder`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/BasicDecoder)
* Obtain the decoder outputs from [`tf.contrib.seq2seq.dynamic_decode`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/dynamic_decode)

In [10]:

def decoding_layer_train(encoder_state, dec_cell, dec_embed_input, 
                         target_sequence_length, max_summary_length, 
                         output_layer, keep_prob):
    """
    Create a decoding layer for training
    :param encoder_state: Encoder State
    :param dec_cell: Decoder RNN Cell
    :param dec_embed_input: Decoder embedded input
    :param target_sequence_length: The lengths of each sequence in the target batch
    :param max_summary_length: The length of the longest sequence in the batch
    :param output_layer: Function to apply the output layer
    :param keep_prob: Dropout keep probability
    :return: BasicDecoderOutput containing training logits and sample_id
    """
    # TODO: Implement Function
    # 2. Construct the decoder cell
    # Helper for the training process. Used by BasicDecoder to read inputs.
    training_helper = tf.contrib.seq2seq.TrainingHelper(inputs=dec_embed_input,
                                                        sequence_length=target_sequence_length,
                                                        time_major=False)
    # Basic decoder
    training_decoder = tf.contrib.seq2seq.BasicDecoder(dec_cell,
                                                       training_helper,
                                                       encoder_state,
                                                       output_layer) 

    # Perform dynamic decoding using the decoder
    training_decoder_output = tf.contrib.seq2seq.dynamic_decode(training_decoder,
                                                                   impute_finished=True,
                                                                   maximum_iterations=max_summary_length)[0]

    return training_decoder_output



"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer_train(decoding_layer_train)

Tests Passed


### Decoding - Inference
Create inference decoder:
* Create a [`tf.contrib.seq2seq.GreedyEmbeddingHelper`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/GreedyEmbeddingHelper)
* Create a [`tf.contrib.seq2seq.BasicDecoder`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/BasicDecoder)
* Obtain the decoder outputs from [`tf.contrib.seq2seq.dynamic_decode`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/dynamic_decode)

In [11]:
def decoding_layer_infer(encoder_state, dec_cell, dec_embeddings, start_of_sequence_id,
                         end_of_sequence_id, max_target_sequence_length,
                         vocab_size, output_layer, batch_size, keep_prob):
    """
    Create a decoding layer for inference
    :param encoder_state: Encoder state
    :param dec_cell: Decoder RNN Cell
    :param dec_embeddings: Decoder embeddings
    :param start_of_sequence_id: GO ID
    :param end_of_sequence_id: EOS Id
    :param max_target_sequence_length: Maximum length of target sequences
    :param vocab_size: Size of decoder/target vocabulary
    :param decoding_scope: TenorFlow Variable Scope for decoding
    :param output_layer: Function to apply the output layer
    :param batch_size: Batch size
    :param keep_prob: Dropout keep probability
    :return: BasicDecoderOutput containing inference logits and sample_id
    """
    # TODO: Implement Function
    
    start_tokens = tf.tile(tf.constant([start_of_sequence_id], dtype=tf.int32), [batch_size], name='start_tokens')

    # Helper for the inference process.
    inference_helper = tf.contrib.seq2seq.GreedyEmbeddingHelper(dec_embeddings,
                                                            start_tokens,
                                                            end_of_sequence_id)

    # Basic decoder
    inference_decoder = tf.contrib.seq2seq.BasicDecoder(dec_cell,
                                                    inference_helper,
                                                    encoder_state,
                                                    output_layer)

    # Perform dynamic decoding using the decoder
    inference_decoder_output = tf.contrib.seq2seq.dynamic_decode(inference_decoder,
                                                        impute_finished=True,
                                                        maximum_iterations=max_target_sequence_length)[0]


    
    return inference_decoder_output



"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer_infer(decoding_layer_infer)

Tests Passed


### Build the Decoding Layer
Implement `decoding_layer()` to create a Decoder RNN layer.

* Embed the target sequences
* Construct the decoder LSTM cell (just like you constructed the encoder cell above)
* Create an output layer to map the outputs of the decoder to the elements of our vocabulary
* Use the your `decoding_layer_train(encoder_state, dec_cell, dec_embed_input, target_sequence_length, max_target_sequence_length, output_layer, keep_prob)` function to get the training logits.
* Use your `decoding_layer_infer(encoder_state, dec_cell, dec_embeddings, start_of_sequence_id, end_of_sequence_id, max_target_sequence_length, vocab_size, output_layer, batch_size, keep_prob)` function to get the inference logits.

Note: You'll need to use [tf.variable_scope](https://www.tensorflow.org/api_docs/python/tf/variable_scope) to share variables between training and inference.

In [12]:
def decoding_layer(dec_input, encoder_state,
                   target_sequence_length, max_target_sequence_length,
                   rnn_size,
                   num_layers, target_vocab_to_int, target_vocab_size,
                   batch_size, keep_prob, decoding_embedding_size):
    """
    Create decoding layer
    :param dec_input: Decoder input
    :param encoder_state: Encoder state
    :param target_sequence_length: The lengths of each sequence in the target batch
    :param max_target_sequence_length: Maximum length of target sequences
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :param target_vocab_size: Size of target vocabulary
    :param batch_size: The size of the batch
    :param keep_prob: Dropout keep probability
    :param decoding_embedding_size: Decoding embedding size
    :return: Tuple of (Training BasicDecoderOutput, Inference BasicDecoderOutput)
    """
    # TODO: Implement Function
    # 1. Decoder Embedding
    dec_embeddings = tf.Variable(tf.random_uniform([target_vocab_size, decoding_embedding_size]))
    dec_embed_input = tf.nn.embedding_lookup(dec_embeddings, dec_input)

    # 2. Construct the decoder cell
    def make_cell(rnn_size):
        dec_cell = tf.contrib.rnn.LSTMCell(rnn_size,
                                           initializer=tf.random_uniform_initializer(-0.1, 0.1, seed=2))
        return dec_cell
#     dec_cell = tf.contrib.rnn.MultiRNNCell(tf.contrib.rnn.BasicLSTMCell(rnn_size) for _ in range(num_layers))
    dec_cell = tf.contrib.rnn.MultiRNNCell([make_cell(rnn_size) for _ in range(num_layers)])
     
    # 3. Dense layer to translate the decoder's output at each time 
    # step into a choice from the target vocabulary
    output_layer = Dense(target_vocab_size,
                         kernel_initializer = tf.truncated_normal_initializer(mean = 0.0, stddev=0.1))


    # 4. Set up a training decoder and an inference decoder
    # Training Decoder
    with tf.variable_scope("decode"):
        training_decoder_output = decoding_layer_train(encoder_state, dec_cell, dec_embed_input, target_sequence_length, 
                             max_target_sequence_length, output_layer, keep_prob)
        
    with tf.variable_scope("decode", reuse=True):
        inference_decoder_output = decoding_layer_infer(encoder_state, dec_cell, dec_embeddings, target_vocab_to_int['<GO>'], 
                                                        target_vocab_to_int['<EOS>'], max_target_sequence_length, target_vocab_size, output_layer, batch_size, keep_prob) 
    return training_decoder_output, inference_decoder_output



"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer(decoding_layer)

Tests Passed


### Build the Neural Network
Apply the functions you implemented above to:

- Encode the input using your `encoding_layer(rnn_inputs, rnn_size, num_layers, keep_prob,  source_sequence_length, source_vocab_size, encoding_embedding_size)`.
- Process target data using your `process_decoder_input(target_data, target_vocab_to_int, batch_size)` function.
- Decode the encoded input using your `decoding_layer(dec_input, enc_state, target_sequence_length, max_target_sentence_length, rnn_size, num_layers, target_vocab_to_int, target_vocab_size, batch_size, keep_prob, dec_embedding_size)` function.

In [32]:
def seq2seq_model(input_data, target_data, keep_prob, batch_size,
                  source_sequence_length, target_sequence_length,
                  max_target_sentence_length,
                  source_vocab_size, target_vocab_size,
                  enc_embedding_size, dec_embedding_size,
                  rnn_size, num_layers, target_vocab_to_int):
    """
    Build the Sequence-to-Sequence part of the neural network
    :param input_data: Input placeholder
    :param target_data: Target placeholder
    :param keep_prob: Dropout keep probability placeholder
    :param batch_size: Batch Size
    :param source_sequence_length: Sequence Lengths of source sequences in the batch
    :param target_sequence_length: Sequence Lengths of target sequences in the batch
    :param source_vocab_size: Source vocabulary size
    :param target_vocab_size: Target vocabulary size
    :param enc_embedding_size: Decoder embedding size
    :param dec_embedding_size: Encoder embedding size
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :return: Tuple of (Training BasicDecoderOutput, Inference BasicDecoderOutput)
    """
    # TODO: Implement Function
    _, enc_state = encoding_layer(input_data, rnn_size, num_layers, 
                                  source_sequence_length = source_sequence_length,
                                  source_vocab_size = source_vocab_size, 
                                  keep_prob = keep_prob,
                                  encoding_embedding_size = enc_embedding_size)
    
    # Prepare the target sequences we'll feed to the decoder in training mode
    dec_input = process_decoder_input(target_data, target_vocab_to_int, batch_size)
    
    # Pass encoder state and decoder inputs to the decoders
    training_decoder_output, inference_decoder_output = decoding_layer(dec_input, enc_state, 
                                                                       target_sequence_length, 
                                                                       max_target_sentence_length, 
                                                                       rnn_size, 
                                                                       num_layers, 
                                                                       target_vocab_to_int, 
                                                                       target_vocab_size, 
                                                                       batch_size, 
                                                                       keep_prob, 
                                                                       dec_embedding_size)
    
    return training_decoder_output, inference_decoder_output


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_seq2seq_model(seq2seq_model)

Tests Passed


## Neural Network Training
### Hyperparameters
Tune the following parameters:

- Set `epochs` to the number of epochs.
- Set `batch_size` to the batch size.
- Set `rnn_size` to the size of the RNNs.
- Set `num_layers` to the number of layers.
- Set `encoding_embedding_size` to the size of the embedding for the encoder.
- Set `decoding_embedding_size` to the size of the embedding for the decoder.
- Set `learning_rate` to the learning rate.
- Set `keep_probability` to the Dropout keep probability
- Set `display_step` to state how many steps between each debug output statement

In [33]:
# Number of Epochs
epochs = 60
# Batch Size
batch_size = 128
# RNN Size
rnn_size = 50
# Number of Layers
num_layers = 2
# Embedding Size
encoding_embedding_size = 15
decoding_embedding_size = 15
# Learning Rate
learning_rate = 0.001
# Dropout Keep Probability
keep_probability = 0.75
display_step = 20

### Build the Graph
Build the graph using the neural network you implemented.

In [34]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
save_path = 'checkpoints/dev'
(source_int_text, target_int_text), (source_vocab_to_int, target_vocab_to_int), _ = helper.load_preprocess()
max_target_sentence_length = max([len(sentence) for sentence in source_int_text])

train_graph = tf.Graph()
with train_graph.as_default():
    input_data, targets, lr, keep_prob, target_sequence_length, max_target_sequence_length, source_sequence_length = model_inputs()

    #sequence_length = tf.placeholder_with_default(max_target_sentence_length, None, name='sequence_length')
    input_shape = tf.shape(input_data)

    train_logits, inference_logits = seq2seq_model(tf.reverse(input_data, [-1]),
                                                   targets,
                                                   keep_prob,
                                                   batch_size,
                                                   source_sequence_length,
                                                   target_sequence_length,
                                                   max_target_sequence_length,
                                                   len(source_vocab_to_int),
                                                   len(target_vocab_to_int),
                                                   encoding_embedding_size,
                                                   decoding_embedding_size,
                                                   rnn_size,
                                                   num_layers,
                                                   target_vocab_to_int)


    training_logits = tf.identity(train_logits.rnn_output, name='logits')
    inference_logits = tf.identity(inference_logits.sample_id, name='predictions')

    masks = tf.sequence_mask(target_sequence_length, max_target_sequence_length, dtype=tf.float32, name='masks')

    with tf.name_scope("optimization"):
        # Loss function
        cost = tf.contrib.seq2seq.sequence_loss(
            training_logits,
            targets,
            masks)

        # Optimizer
        optimizer = tf.train.AdamOptimizer(lr)

        # Gradient Clipping
        gradients = optimizer.compute_gradients(cost)
        capped_gradients = [(tf.clip_by_value(grad, -1., 1.), var) for grad, var in gradients if grad is not None]
        train_op = optimizer.apply_gradients(capped_gradients)


Batch and pad the source and target sequences

In [35]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
def pad_sentence_batch(sentence_batch, pad_int):
    """Pad sentences with <PAD> so that each sentence of a batch has the same length"""
    max_sentence = max([len(sentence) for sentence in sentence_batch])
    return [sentence + [pad_int] * (max_sentence - len(sentence)) for sentence in sentence_batch]


def get_batches(sources, targets, batch_size, source_pad_int, target_pad_int):
    """Batch targets, sources, and the lengths of their sentences together"""
    for batch_i in range(0, len(sources)//batch_size):
        start_i = batch_i * batch_size

        # Slice the right amount for the batch
        sources_batch = sources[start_i:start_i + batch_size]
        targets_batch = targets[start_i:start_i + batch_size]

        # Pad
        pad_sources_batch = np.array(pad_sentence_batch(sources_batch, source_pad_int))
        pad_targets_batch = np.array(pad_sentence_batch(targets_batch, target_pad_int))

        # Need the lengths for the _lengths parameters
        pad_targets_lengths = []
        for target in pad_targets_batch:
            pad_targets_lengths.append(len(target))

        pad_source_lengths = []
        for source in pad_sources_batch:
            pad_source_lengths.append(len(source))

        yield pad_sources_batch, pad_targets_batch, pad_source_lengths, pad_targets_lengths


### Train
Train the neural network on the preprocessed data. If you have a hard time getting a good loss, check the forms to see if anyone is having the same problem.

In [36]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
def get_accuracy(target, logits):
    """
    Calculate accuracy
    """
    max_seq = max(target.shape[1], logits.shape[1])
    if max_seq - target.shape[1]:
        target = np.pad(
            target,
            [(0,0),(0,max_seq - target.shape[1])],
            'constant')
    if max_seq - logits.shape[1]:
        logits = np.pad(
            logits,
            [(0,0),(0,max_seq - logits.shape[1])],
            'constant')

    return np.mean(np.equal(target, logits))

# Split data to training and validation sets
train_source = source_int_text[batch_size:]
train_target = target_int_text[batch_size:]
valid_source = source_int_text[:batch_size]
valid_target = target_int_text[:batch_size]
(valid_sources_batch, valid_targets_batch, valid_sources_lengths, valid_targets_lengths ) = next(get_batches(valid_source,
                                                                                                             valid_target,
                                                                                                             batch_size,
                                                                                                             source_vocab_to_int['<PAD>'],
                                                                                                             target_vocab_to_int['<PAD>']))                                                                                                  
with tf.Session(graph=train_graph) as sess:
    sess.run(tf.global_variables_initializer())

    for epoch_i in range(epochs):
        for batch_i, (source_batch, target_batch, sources_lengths, targets_lengths) in enumerate(
                get_batches(train_source, train_target, batch_size,
                            source_vocab_to_int['<PAD>'],
                            target_vocab_to_int['<PAD>'])):

            _, loss = sess.run(
                [train_op, cost],
                {input_data: source_batch,
                 targets: target_batch,
                 lr: learning_rate,
                 target_sequence_length: targets_lengths,
                 source_sequence_length: sources_lengths,
                 keep_prob: keep_probability})


            if batch_i % display_step == 0 and batch_i > 0:


                batch_train_logits = sess.run(
                    inference_logits,
                    {input_data: source_batch,
                     source_sequence_length: sources_lengths,
                     target_sequence_length: targets_lengths,
                     keep_prob: 1.0})


                batch_valid_logits = sess.run(
                    inference_logits,
                    {input_data: valid_sources_batch,
                     source_sequence_length: valid_sources_lengths,
                     target_sequence_length: valid_targets_lengths,
                     keep_prob: 1.0})

                train_acc = get_accuracy(target_batch, batch_train_logits)

                valid_acc = get_accuracy(valid_targets_batch, batch_valid_logits)

                print('Epoch {:>3} Batch {:>4}/{} - Train Accuracy: {:>6.4f}, Validation Accuracy: {:>6.4f}, Loss: {:>6.4f}'
                      .format(epoch_i, batch_i, len(source_int_text) // batch_size, train_acc, valid_acc, loss))

    # Save Model
    saver = tf.train.Saver()
    saver.save(sess, save_path)
    print('Model Trained and Saved')

Epoch   0 Batch   20/1077 - Train Accuracy: 0.2359, Validation Accuracy: 0.3089, Loss: 4.7571
Epoch   0 Batch   40/1077 - Train Accuracy: 0.2406, Validation Accuracy: 0.3161, Loss: 3.7640
Epoch   0 Batch   60/1077 - Train Accuracy: 0.2972, Validation Accuracy: 0.3363, Loss: 3.3303
Epoch   0 Batch   80/1077 - Train Accuracy: 0.2715, Validation Accuracy: 0.3363, Loss: 3.2839
Epoch   0 Batch  100/1077 - Train Accuracy: 0.2617, Validation Accuracy: 0.3363, Loss: 3.2375
Epoch   0 Batch  120/1077 - Train Accuracy: 0.3000, Validation Accuracy: 0.3622, Loss: 3.1316
Epoch   0 Batch  140/1077 - Train Accuracy: 0.2738, Validation Accuracy: 0.3668, Loss: 3.2006
Epoch   0 Batch  160/1077 - Train Accuracy: 0.3156, Validation Accuracy: 0.3704, Loss: 2.9518
Epoch   0 Batch  180/1077 - Train Accuracy: 0.3289, Validation Accuracy: 0.3860, Loss: 2.9203
Epoch   0 Batch  200/1077 - Train Accuracy: 0.3488, Validation Accuracy: 0.4031, Loss: 2.8580
Epoch   0 Batch  220/1077 - Train Accuracy: 0.3363, Validati

Epoch   1 Batch  720/1077 - Train Accuracy: 0.4342, Validation Accuracy: 0.4840, Loss: 1.0432
Epoch   1 Batch  740/1077 - Train Accuracy: 0.4699, Validation Accuracy: 0.4787, Loss: 0.9546
Epoch   1 Batch  760/1077 - Train Accuracy: 0.4520, Validation Accuracy: 0.4769, Loss: 0.9893
Epoch   1 Batch  780/1077 - Train Accuracy: 0.4332, Validation Accuracy: 0.4787, Loss: 0.9879
Epoch   1 Batch  800/1077 - Train Accuracy: 0.4676, Validation Accuracy: 0.4968, Loss: 0.9698
Epoch   1 Batch  820/1077 - Train Accuracy: 0.3926, Validation Accuracy: 0.4865, Loss: 0.9626
Epoch   1 Batch  840/1077 - Train Accuracy: 0.4355, Validation Accuracy: 0.4805, Loss: 0.9231
Epoch   1 Batch  860/1077 - Train Accuracy: 0.4628, Validation Accuracy: 0.4840, Loss: 0.9053
Epoch   1 Batch  880/1077 - Train Accuracy: 0.4633, Validation Accuracy: 0.4936, Loss: 0.8920
Epoch   1 Batch  900/1077 - Train Accuracy: 0.4566, Validation Accuracy: 0.4776, Loss: 0.9485
Epoch   1 Batch  920/1077 - Train Accuracy: 0.4219, Validati

Epoch   3 Batch  360/1077 - Train Accuracy: 0.5895, Validation Accuracy: 0.6140, Loss: 0.6212
Epoch   3 Batch  380/1077 - Train Accuracy: 0.5879, Validation Accuracy: 0.6374, Loss: 0.5994
Epoch   3 Batch  400/1077 - Train Accuracy: 0.6062, Validation Accuracy: 0.6158, Loss: 0.6187
Epoch   3 Batch  420/1077 - Train Accuracy: 0.6020, Validation Accuracy: 0.6243, Loss: 0.5849
Epoch   3 Batch  440/1077 - Train Accuracy: 0.5848, Validation Accuracy: 0.6094, Loss: 0.6219
Epoch   3 Batch  460/1077 - Train Accuracy: 0.6027, Validation Accuracy: 0.6122, Loss: 0.6159
Epoch   3 Batch  480/1077 - Train Accuracy: 0.6197, Validation Accuracy: 0.6317, Loss: 0.6103
Epoch   3 Batch  500/1077 - Train Accuracy: 0.6289, Validation Accuracy: 0.6346, Loss: 0.5981
Epoch   3 Batch  520/1077 - Train Accuracy: 0.6570, Validation Accuracy: 0.6360, Loss: 0.5554
Epoch   3 Batch  540/1077 - Train Accuracy: 0.6234, Validation Accuracy: 0.6165, Loss: 0.5587
Epoch   3 Batch  560/1077 - Train Accuracy: 0.6211, Validati

Epoch   4 Batch 1060/1077 - Train Accuracy: 0.6863, Validation Accuracy: 0.6616, Loss: 0.4123
Epoch   5 Batch   20/1077 - Train Accuracy: 0.6922, Validation Accuracy: 0.6534, Loss: 0.4065
Epoch   5 Batch   40/1077 - Train Accuracy: 0.6844, Validation Accuracy: 0.6669, Loss: 0.4249
Epoch   5 Batch   60/1077 - Train Accuracy: 0.7091, Validation Accuracy: 0.6385, Loss: 0.3960
Epoch   5 Batch   80/1077 - Train Accuracy: 0.6887, Validation Accuracy: 0.6648, Loss: 0.4189
Epoch   5 Batch  100/1077 - Train Accuracy: 0.7082, Validation Accuracy: 0.6619, Loss: 0.4168
Epoch   5 Batch  120/1077 - Train Accuracy: 0.6813, Validation Accuracy: 0.6722, Loss: 0.4314
Epoch   5 Batch  140/1077 - Train Accuracy: 0.6768, Validation Accuracy: 0.6857, Loss: 0.4260
Epoch   5 Batch  160/1077 - Train Accuracy: 0.6980, Validation Accuracy: 0.6687, Loss: 0.4060
Epoch   5 Batch  180/1077 - Train Accuracy: 0.6902, Validation Accuracy: 0.6690, Loss: 0.3928
Epoch   5 Batch  200/1077 - Train Accuracy: 0.6727, Validati

Epoch   6 Batch  700/1077 - Train Accuracy: 0.7344, Validation Accuracy: 0.7081, Loss: 0.2723
Epoch   6 Batch  720/1077 - Train Accuracy: 0.7467, Validation Accuracy: 0.7244, Loss: 0.3138
Epoch   6 Batch  740/1077 - Train Accuracy: 0.7559, Validation Accuracy: 0.7241, Loss: 0.2840
Epoch   6 Batch  760/1077 - Train Accuracy: 0.7484, Validation Accuracy: 0.7362, Loss: 0.3061
Epoch   6 Batch  780/1077 - Train Accuracy: 0.7480, Validation Accuracy: 0.7468, Loss: 0.3097
Epoch   6 Batch  800/1077 - Train Accuracy: 0.7637, Validation Accuracy: 0.7184, Loss: 0.2845
Epoch   6 Batch  820/1077 - Train Accuracy: 0.7027, Validation Accuracy: 0.7241, Loss: 0.3014
Epoch   6 Batch  840/1077 - Train Accuracy: 0.7703, Validation Accuracy: 0.7330, Loss: 0.2714
Epoch   6 Batch  860/1077 - Train Accuracy: 0.7403, Validation Accuracy: 0.7376, Loss: 0.2879
Epoch   6 Batch  880/1077 - Train Accuracy: 0.7887, Validation Accuracy: 0.7340, Loss: 0.2728
Epoch   6 Batch  900/1077 - Train Accuracy: 0.7957, Validati

Epoch   8 Batch  340/1077 - Train Accuracy: 0.8438, Validation Accuracy: 0.7976, Loss: 0.1963
Epoch   8 Batch  360/1077 - Train Accuracy: 0.8418, Validation Accuracy: 0.7901, Loss: 0.1902
Epoch   8 Batch  380/1077 - Train Accuracy: 0.8203, Validation Accuracy: 0.7887, Loss: 0.1926
Epoch   8 Batch  400/1077 - Train Accuracy: 0.8047, Validation Accuracy: 0.7749, Loss: 0.2179
Epoch   8 Batch  420/1077 - Train Accuracy: 0.8770, Validation Accuracy: 0.7933, Loss: 0.1774
Epoch   8 Batch  440/1077 - Train Accuracy: 0.7992, Validation Accuracy: 0.7852, Loss: 0.2071
Epoch   8 Batch  460/1077 - Train Accuracy: 0.8180, Validation Accuracy: 0.7905, Loss: 0.2057
Epoch   8 Batch  480/1077 - Train Accuracy: 0.8363, Validation Accuracy: 0.7955, Loss: 0.2006
Epoch   8 Batch  500/1077 - Train Accuracy: 0.8473, Validation Accuracy: 0.7887, Loss: 0.1870
Epoch   8 Batch  520/1077 - Train Accuracy: 0.8467, Validation Accuracy: 0.7880, Loss: 0.1712
Epoch   8 Batch  540/1077 - Train Accuracy: 0.8070, Validati

Epoch   9 Batch 1040/1077 - Train Accuracy: 0.8725, Validation Accuracy: 0.8253, Loss: 0.1448
Epoch   9 Batch 1060/1077 - Train Accuracy: 0.8551, Validation Accuracy: 0.8313, Loss: 0.1375
Epoch  10 Batch   20/1077 - Train Accuracy: 0.8473, Validation Accuracy: 0.8260, Loss: 0.1184
Epoch  10 Batch   40/1077 - Train Accuracy: 0.8934, Validation Accuracy: 0.8473, Loss: 0.1299
Epoch  10 Batch   60/1077 - Train Accuracy: 0.8452, Validation Accuracy: 0.8267, Loss: 0.1269
Epoch  10 Batch   80/1077 - Train Accuracy: 0.8820, Validation Accuracy: 0.8406, Loss: 0.1356
Epoch  10 Batch  100/1077 - Train Accuracy: 0.8645, Validation Accuracy: 0.8423, Loss: 0.1330
Epoch  10 Batch  120/1077 - Train Accuracy: 0.8809, Validation Accuracy: 0.8370, Loss: 0.1560
Epoch  10 Batch  140/1077 - Train Accuracy: 0.8684, Validation Accuracy: 0.8203, Loss: 0.1351
Epoch  10 Batch  160/1077 - Train Accuracy: 0.8812, Validation Accuracy: 0.8427, Loss: 0.1319
Epoch  10 Batch  180/1077 - Train Accuracy: 0.8625, Validati

Epoch  11 Batch  680/1077 - Train Accuracy: 0.8668, Validation Accuracy: 0.8363, Loss: 0.1065
Epoch  11 Batch  700/1077 - Train Accuracy: 0.8793, Validation Accuracy: 0.8494, Loss: 0.0839
Epoch  11 Batch  720/1077 - Train Accuracy: 0.8688, Validation Accuracy: 0.8459, Loss: 0.1129
Epoch  11 Batch  740/1077 - Train Accuracy: 0.8590, Validation Accuracy: 0.8480, Loss: 0.0928
Epoch  11 Batch  760/1077 - Train Accuracy: 0.8977, Validation Accuracy: 0.8487, Loss: 0.1071
Epoch  11 Batch  780/1077 - Train Accuracy: 0.8617, Validation Accuracy: 0.8537, Loss: 0.1156
Epoch  11 Batch  800/1077 - Train Accuracy: 0.8992, Validation Accuracy: 0.8530, Loss: 0.0979
Epoch  11 Batch  820/1077 - Train Accuracy: 0.8613, Validation Accuracy: 0.8580, Loss: 0.1080
Epoch  11 Batch  840/1077 - Train Accuracy: 0.8750, Validation Accuracy: 0.8665, Loss: 0.0904
Epoch  11 Batch  860/1077 - Train Accuracy: 0.8743, Validation Accuracy: 0.8633, Loss: 0.1141
Epoch  11 Batch  880/1077 - Train Accuracy: 0.9020, Validati

Epoch  13 Batch  320/1077 - Train Accuracy: 0.9148, Validation Accuracy: 0.8743, Loss: 0.0868
Epoch  13 Batch  340/1077 - Train Accuracy: 0.9354, Validation Accuracy: 0.8494, Loss: 0.0702
Epoch  13 Batch  360/1077 - Train Accuracy: 0.9223, Validation Accuracy: 0.8661, Loss: 0.0641
Epoch  13 Batch  380/1077 - Train Accuracy: 0.9246, Validation Accuracy: 0.8732, Loss: 0.0663
Epoch  13 Batch  400/1077 - Train Accuracy: 0.8844, Validation Accuracy: 0.8700, Loss: 0.0862
Epoch  13 Batch  420/1077 - Train Accuracy: 0.9223, Validation Accuracy: 0.8604, Loss: 0.0646
Epoch  13 Batch  440/1077 - Train Accuracy: 0.8680, Validation Accuracy: 0.8675, Loss: 0.0868
Epoch  13 Batch  460/1077 - Train Accuracy: 0.8887, Validation Accuracy: 0.8473, Loss: 0.0845
Epoch  13 Batch  480/1077 - Train Accuracy: 0.9075, Validation Accuracy: 0.8729, Loss: 0.0758
Epoch  13 Batch  500/1077 - Train Accuracy: 0.9187, Validation Accuracy: 0.8903, Loss: 0.0696
Epoch  13 Batch  520/1077 - Train Accuracy: 0.9360, Validati

Epoch  14 Batch 1020/1077 - Train Accuracy: 0.9426, Validation Accuracy: 0.8864, Loss: 0.0523
Epoch  14 Batch 1040/1077 - Train Accuracy: 0.9206, Validation Accuracy: 0.8860, Loss: 0.0622
Epoch  14 Batch 1060/1077 - Train Accuracy: 0.9082, Validation Accuracy: 0.8963, Loss: 0.0567
Epoch  15 Batch   20/1077 - Train Accuracy: 0.8922, Validation Accuracy: 0.9102, Loss: 0.0510
Epoch  15 Batch   40/1077 - Train Accuracy: 0.9227, Validation Accuracy: 0.8881, Loss: 0.0567
Epoch  15 Batch   60/1077 - Train Accuracy: 0.9029, Validation Accuracy: 0.8960, Loss: 0.0532
Epoch  15 Batch   80/1077 - Train Accuracy: 0.9141, Validation Accuracy: 0.8846, Loss: 0.0606
Epoch  15 Batch  100/1077 - Train Accuracy: 0.9152, Validation Accuracy: 0.8984, Loss: 0.0566
Epoch  15 Batch  120/1077 - Train Accuracy: 0.9227, Validation Accuracy: 0.9134, Loss: 0.0702
Epoch  15 Batch  140/1077 - Train Accuracy: 0.9120, Validation Accuracy: 0.8867, Loss: 0.0558
Epoch  15 Batch  160/1077 - Train Accuracy: 0.9078, Validati

Epoch  16 Batch  660/1077 - Train Accuracy: 0.9012, Validation Accuracy: 0.9222, Loss: 0.0520
Epoch  16 Batch  680/1077 - Train Accuracy: 0.9219, Validation Accuracy: 0.9169, Loss: 0.0640
Epoch  16 Batch  700/1077 - Train Accuracy: 0.9352, Validation Accuracy: 0.9219, Loss: 0.0413
Epoch  16 Batch  720/1077 - Train Accuracy: 0.9046, Validation Accuracy: 0.9194, Loss: 0.0657
Epoch  16 Batch  740/1077 - Train Accuracy: 0.9242, Validation Accuracy: 0.9141, Loss: 0.0452
Epoch  16 Batch  760/1077 - Train Accuracy: 0.9305, Validation Accuracy: 0.9116, Loss: 0.0596
Epoch  16 Batch  780/1077 - Train Accuracy: 0.8934, Validation Accuracy: 0.9116, Loss: 0.0638
Epoch  16 Batch  800/1077 - Train Accuracy: 0.9105, Validation Accuracy: 0.9208, Loss: 0.0519
Epoch  16 Batch  820/1077 - Train Accuracy: 0.8910, Validation Accuracy: 0.9208, Loss: 0.0554
Epoch  16 Batch  840/1077 - Train Accuracy: 0.9121, Validation Accuracy: 0.9197, Loss: 0.0521
Epoch  16 Batch  860/1077 - Train Accuracy: 0.9282, Validati

Epoch  18 Batch  300/1077 - Train Accuracy: 0.9482, Validation Accuracy: 0.9279, Loss: 0.0399
Epoch  18 Batch  320/1077 - Train Accuracy: 0.9313, Validation Accuracy: 0.9176, Loss: 0.0572
Epoch  18 Batch  340/1077 - Train Accuracy: 0.9564, Validation Accuracy: 0.9183, Loss: 0.0429
Epoch  18 Batch  360/1077 - Train Accuracy: 0.9367, Validation Accuracy: 0.9283, Loss: 0.0354
Epoch  18 Batch  380/1077 - Train Accuracy: 0.9539, Validation Accuracy: 0.9283, Loss: 0.0377
Epoch  18 Batch  400/1077 - Train Accuracy: 0.9320, Validation Accuracy: 0.9219, Loss: 0.0518
Epoch  18 Batch  420/1077 - Train Accuracy: 0.9461, Validation Accuracy: 0.9226, Loss: 0.0350
Epoch  18 Batch  440/1077 - Train Accuracy: 0.9094, Validation Accuracy: 0.9279, Loss: 0.0532
Epoch  18 Batch  460/1077 - Train Accuracy: 0.9320, Validation Accuracy: 0.9197, Loss: 0.0534
Epoch  18 Batch  480/1077 - Train Accuracy: 0.9330, Validation Accuracy: 0.9386, Loss: 0.0439
Epoch  18 Batch  500/1077 - Train Accuracy: 0.9289, Validati

Epoch  19 Batch 1000/1077 - Train Accuracy: 0.9304, Validation Accuracy: 0.9190, Loss: 0.0401
Epoch  19 Batch 1020/1077 - Train Accuracy: 0.9520, Validation Accuracy: 0.9371, Loss: 0.0323
Epoch  19 Batch 1040/1077 - Train Accuracy: 0.9379, Validation Accuracy: 0.9332, Loss: 0.0379
Epoch  19 Batch 1060/1077 - Train Accuracy: 0.9262, Validation Accuracy: 0.9371, Loss: 0.0381
Epoch  20 Batch   20/1077 - Train Accuracy: 0.9293, Validation Accuracy: 0.9450, Loss: 0.0353
Epoch  20 Batch   40/1077 - Train Accuracy: 0.9414, Validation Accuracy: 0.9180, Loss: 0.0455
Epoch  20 Batch   60/1077 - Train Accuracy: 0.9401, Validation Accuracy: 0.9293, Loss: 0.0339
Epoch  20 Batch   80/1077 - Train Accuracy: 0.9246, Validation Accuracy: 0.9347, Loss: 0.0385
Epoch  20 Batch  100/1077 - Train Accuracy: 0.9363, Validation Accuracy: 0.9553, Loss: 0.0407
Epoch  20 Batch  120/1077 - Train Accuracy: 0.9414, Validation Accuracy: 0.9162, Loss: 0.0479
Epoch  20 Batch  140/1077 - Train Accuracy: 0.9428, Validati

Epoch  21 Batch  640/1077 - Train Accuracy: 0.9178, Validation Accuracy: 0.9478, Loss: 0.0363
Epoch  21 Batch  660/1077 - Train Accuracy: 0.9324, Validation Accuracy: 0.9485, Loss: 0.0335
Epoch  21 Batch  680/1077 - Train Accuracy: 0.9275, Validation Accuracy: 0.9371, Loss: 0.0423
Epoch  21 Batch  700/1077 - Train Accuracy: 0.9613, Validation Accuracy: 0.9386, Loss: 0.0272
Epoch  21 Batch  720/1077 - Train Accuracy: 0.9252, Validation Accuracy: 0.9471, Loss: 0.0461
Epoch  21 Batch  740/1077 - Train Accuracy: 0.9391, Validation Accuracy: 0.9620, Loss: 0.0315
Epoch  21 Batch  760/1077 - Train Accuracy: 0.9578, Validation Accuracy: 0.9425, Loss: 0.0401
Epoch  21 Batch  780/1077 - Train Accuracy: 0.9391, Validation Accuracy: 0.9442, Loss: 0.0439
Epoch  21 Batch  800/1077 - Train Accuracy: 0.9336, Validation Accuracy: 0.9464, Loss: 0.0377
Epoch  21 Batch  820/1077 - Train Accuracy: 0.9426, Validation Accuracy: 0.9368, Loss: 0.0366
Epoch  21 Batch  840/1077 - Train Accuracy: 0.9520, Validati

Epoch  23 Batch  280/1077 - Train Accuracy: 0.9137, Validation Accuracy: 0.9510, Loss: 0.0357
Epoch  23 Batch  300/1077 - Train Accuracy: 0.9539, Validation Accuracy: 0.9460, Loss: 0.0261
Epoch  23 Batch  320/1077 - Train Accuracy: 0.9332, Validation Accuracy: 0.9343, Loss: 0.0466
Epoch  23 Batch  340/1077 - Train Accuracy: 0.9597, Validation Accuracy: 0.9389, Loss: 0.0320
Epoch  23 Batch  360/1077 - Train Accuracy: 0.9414, Validation Accuracy: 0.9361, Loss: 0.0257
Epoch  23 Batch  380/1077 - Train Accuracy: 0.9695, Validation Accuracy: 0.9435, Loss: 0.0262
Epoch  23 Batch  400/1077 - Train Accuracy: 0.9363, Validation Accuracy: 0.9375, Loss: 0.0372
Epoch  23 Batch  420/1077 - Train Accuracy: 0.9723, Validation Accuracy: 0.9560, Loss: 0.0233
Epoch  23 Batch  440/1077 - Train Accuracy: 0.9371, Validation Accuracy: 0.9308, Loss: 0.0383
Epoch  23 Batch  460/1077 - Train Accuracy: 0.9477, Validation Accuracy: 0.9368, Loss: 0.0395
Epoch  23 Batch  480/1077 - Train Accuracy: 0.9634, Validati

Epoch  24 Batch  980/1077 - Train Accuracy: 0.9473, Validation Accuracy: 0.9563, Loss: 0.0334
Epoch  24 Batch 1000/1077 - Train Accuracy: 0.9438, Validation Accuracy: 0.9450, Loss: 0.0327
Epoch  24 Batch 1020/1077 - Train Accuracy: 0.9684, Validation Accuracy: 0.9457, Loss: 0.0204
Epoch  24 Batch 1040/1077 - Train Accuracy: 0.9527, Validation Accuracy: 0.9506, Loss: 0.0290
Epoch  24 Batch 1060/1077 - Train Accuracy: 0.9375, Validation Accuracy: 0.9499, Loss: 0.0277
Epoch  25 Batch   20/1077 - Train Accuracy: 0.9539, Validation Accuracy: 0.9510, Loss: 0.0253
Epoch  25 Batch   40/1077 - Train Accuracy: 0.9660, Validation Accuracy: 0.9460, Loss: 0.0292
Epoch  25 Batch   60/1077 - Train Accuracy: 0.9513, Validation Accuracy: 0.9482, Loss: 0.0233
Epoch  25 Batch   80/1077 - Train Accuracy: 0.9516, Validation Accuracy: 0.9560, Loss: 0.0302
Epoch  25 Batch  100/1077 - Train Accuracy: 0.9535, Validation Accuracy: 0.9485, Loss: 0.0282
Epoch  25 Batch  120/1077 - Train Accuracy: 0.9441, Validati

Epoch  26 Batch  620/1077 - Train Accuracy: 0.9766, Validation Accuracy: 0.9638, Loss: 0.0238
Epoch  26 Batch  640/1077 - Train Accuracy: 0.9390, Validation Accuracy: 0.9695, Loss: 0.0298
Epoch  26 Batch  660/1077 - Train Accuracy: 0.9668, Validation Accuracy: 0.9592, Loss: 0.0222
Epoch  26 Batch  680/1077 - Train Accuracy: 0.9364, Validation Accuracy: 0.9563, Loss: 0.0333
Epoch  26 Batch  700/1077 - Train Accuracy: 0.9738, Validation Accuracy: 0.9474, Loss: 0.0213
Epoch  26 Batch  720/1077 - Train Accuracy: 0.9428, Validation Accuracy: 0.9553, Loss: 0.0353
Epoch  26 Batch  740/1077 - Train Accuracy: 0.9391, Validation Accuracy: 0.9411, Loss: 0.0266
Epoch  26 Batch  760/1077 - Train Accuracy: 0.9621, Validation Accuracy: 0.9542, Loss: 0.0309
Epoch  26 Batch  780/1077 - Train Accuracy: 0.9430, Validation Accuracy: 0.9599, Loss: 0.0364
Epoch  26 Batch  800/1077 - Train Accuracy: 0.9441, Validation Accuracy: 0.9645, Loss: 0.0308
Epoch  26 Batch  820/1077 - Train Accuracy: 0.9672, Validati

Epoch  28 Batch  260/1077 - Train Accuracy: 0.9580, Validation Accuracy: 0.9560, Loss: 0.0224
Epoch  28 Batch  280/1077 - Train Accuracy: 0.9344, Validation Accuracy: 0.9457, Loss: 0.0283
Epoch  28 Batch  300/1077 - Train Accuracy: 0.9630, Validation Accuracy: 0.9251, Loss: 0.0232
Epoch  28 Batch  320/1077 - Train Accuracy: 0.9496, Validation Accuracy: 0.9322, Loss: 0.0406
Epoch  28 Batch  340/1077 - Train Accuracy: 0.9544, Validation Accuracy: 0.9478, Loss: 0.0273
Epoch  28 Batch  360/1077 - Train Accuracy: 0.9680, Validation Accuracy: 0.9499, Loss: 0.0194
Epoch  28 Batch  380/1077 - Train Accuracy: 0.9797, Validation Accuracy: 0.9389, Loss: 0.0223
Epoch  28 Batch  400/1077 - Train Accuracy: 0.9461, Validation Accuracy: 0.9482, Loss: 0.0314
Epoch  28 Batch  420/1077 - Train Accuracy: 0.9871, Validation Accuracy: 0.9567, Loss: 0.0180
Epoch  28 Batch  440/1077 - Train Accuracy: 0.9570, Validation Accuracy: 0.9521, Loss: 0.0290
Epoch  28 Batch  460/1077 - Train Accuracy: 0.9410, Validati

Epoch  29 Batch  960/1077 - Train Accuracy: 0.9624, Validation Accuracy: 0.9510, Loss: 0.0260
Epoch  29 Batch  980/1077 - Train Accuracy: 0.9406, Validation Accuracy: 0.9588, Loss: 0.0315
Epoch  29 Batch 1000/1077 - Train Accuracy: 0.9449, Validation Accuracy: 0.9496, Loss: 0.0308
Epoch  29 Batch 1020/1077 - Train Accuracy: 0.9680, Validation Accuracy: 0.9521, Loss: 0.0160
Epoch  29 Batch 1040/1077 - Train Accuracy: 0.9552, Validation Accuracy: 0.9549, Loss: 0.0242
Epoch  29 Batch 1060/1077 - Train Accuracy: 0.9586, Validation Accuracy: 0.9688, Loss: 0.0220
Epoch  30 Batch   20/1077 - Train Accuracy: 0.9547, Validation Accuracy: 0.9613, Loss: 0.0221
Epoch  30 Batch   40/1077 - Train Accuracy: 0.9805, Validation Accuracy: 0.9577, Loss: 0.0222
Epoch  30 Batch   60/1077 - Train Accuracy: 0.9565, Validation Accuracy: 0.9403, Loss: 0.0203
Epoch  30 Batch   80/1077 - Train Accuracy: 0.9574, Validation Accuracy: 0.9556, Loss: 0.0250
Epoch  30 Batch  100/1077 - Train Accuracy: 0.9688, Validati

Epoch  31 Batch  600/1077 - Train Accuracy: 0.9784, Validation Accuracy: 0.9634, Loss: 0.0253
Epoch  31 Batch  620/1077 - Train Accuracy: 0.9734, Validation Accuracy: 0.9709, Loss: 0.0206
Epoch  31 Batch  640/1077 - Train Accuracy: 0.9751, Validation Accuracy: 0.9670, Loss: 0.0276
Epoch  31 Batch  660/1077 - Train Accuracy: 0.9828, Validation Accuracy: 0.9531, Loss: 0.0173
Epoch  31 Batch  680/1077 - Train Accuracy: 0.9446, Validation Accuracy: 0.9602, Loss: 0.0254
Epoch  31 Batch  700/1077 - Train Accuracy: 0.9844, Validation Accuracy: 0.9581, Loss: 0.0177
Epoch  31 Batch  720/1077 - Train Accuracy: 0.9535, Validation Accuracy: 0.9474, Loss: 0.0285
Epoch  31 Batch  740/1077 - Train Accuracy: 0.9313, Validation Accuracy: 0.9489, Loss: 0.0218
Epoch  31 Batch  760/1077 - Train Accuracy: 0.9664, Validation Accuracy: 0.9446, Loss: 0.0223
Epoch  31 Batch  780/1077 - Train Accuracy: 0.9488, Validation Accuracy: 0.9638, Loss: 0.0341
Epoch  31 Batch  800/1077 - Train Accuracy: 0.9594, Validati

Epoch  33 Batch  240/1077 - Train Accuracy: 0.9832, Validation Accuracy: 0.9567, Loss: 0.0182
Epoch  33 Batch  260/1077 - Train Accuracy: 0.9639, Validation Accuracy: 0.9663, Loss: 0.0184
Epoch  33 Batch  280/1077 - Train Accuracy: 0.9414, Validation Accuracy: 0.9435, Loss: 0.0250
Epoch  33 Batch  300/1077 - Train Accuracy: 0.9704, Validation Accuracy: 0.9482, Loss: 0.0208
Epoch  33 Batch  320/1077 - Train Accuracy: 0.9453, Validation Accuracy: 0.9439, Loss: 0.0348
Epoch  33 Batch  340/1077 - Train Accuracy: 0.9667, Validation Accuracy: 0.9602, Loss: 0.0240
Epoch  33 Batch  360/1077 - Train Accuracy: 0.9707, Validation Accuracy: 0.9499, Loss: 0.0153
Epoch  33 Batch  380/1077 - Train Accuracy: 0.9848, Validation Accuracy: 0.9538, Loss: 0.0167
Epoch  33 Batch  400/1077 - Train Accuracy: 0.9477, Validation Accuracy: 0.9542, Loss: 0.0266
Epoch  33 Batch  420/1077 - Train Accuracy: 0.9898, Validation Accuracy: 0.9769, Loss: 0.0149
Epoch  33 Batch  440/1077 - Train Accuracy: 0.9578, Validati

Epoch  34 Batch  940/1077 - Train Accuracy: 0.9598, Validation Accuracy: 0.9634, Loss: 0.0194
Epoch  34 Batch  960/1077 - Train Accuracy: 0.9639, Validation Accuracy: 0.9574, Loss: 0.0196
Epoch  34 Batch  980/1077 - Train Accuracy: 0.9648, Validation Accuracy: 0.9673, Loss: 0.0220
Epoch  34 Batch 1000/1077 - Train Accuracy: 0.9554, Validation Accuracy: 0.9585, Loss: 0.0261
Epoch  34 Batch 1020/1077 - Train Accuracy: 0.9898, Validation Accuracy: 0.9478, Loss: 0.0106
Epoch  34 Batch 1040/1077 - Train Accuracy: 0.9737, Validation Accuracy: 0.9592, Loss: 0.0192
Epoch  34 Batch 1060/1077 - Train Accuracy: 0.9723, Validation Accuracy: 0.9688, Loss: 0.0161
Epoch  35 Batch   20/1077 - Train Accuracy: 0.9559, Validation Accuracy: 0.9641, Loss: 0.0197
Epoch  35 Batch   40/1077 - Train Accuracy: 0.9938, Validation Accuracy: 0.9570, Loss: 0.0184
Epoch  35 Batch   60/1077 - Train Accuracy: 0.9557, Validation Accuracy: 0.9599, Loss: 0.0171
Epoch  35 Batch   80/1077 - Train Accuracy: 0.9551, Validati

Epoch  36 Batch  580/1077 - Train Accuracy: 0.9594, Validation Accuracy: 0.9585, Loss: 0.0159
Epoch  36 Batch  600/1077 - Train Accuracy: 0.9743, Validation Accuracy: 0.9652, Loss: 0.0225
Epoch  36 Batch  620/1077 - Train Accuracy: 0.9742, Validation Accuracy: 0.9702, Loss: 0.0186
Epoch  36 Batch  640/1077 - Train Accuracy: 0.9710, Validation Accuracy: 0.9599, Loss: 0.0239
Epoch  36 Batch  660/1077 - Train Accuracy: 0.9812, Validation Accuracy: 0.9680, Loss: 0.0138
Epoch  36 Batch  680/1077 - Train Accuracy: 0.9472, Validation Accuracy: 0.9769, Loss: 0.0215
Epoch  36 Batch  700/1077 - Train Accuracy: 0.9855, Validation Accuracy: 0.9599, Loss: 0.0153
Epoch  36 Batch  720/1077 - Train Accuracy: 0.9634, Validation Accuracy: 0.9702, Loss: 0.0267
Epoch  36 Batch  740/1077 - Train Accuracy: 0.9523, Validation Accuracy: 0.9677, Loss: 0.0181
Epoch  36 Batch  760/1077 - Train Accuracy: 0.9738, Validation Accuracy: 0.9577, Loss: 0.0179
Epoch  36 Batch  780/1077 - Train Accuracy: 0.9602, Validati

Epoch  38 Batch  220/1077 - Train Accuracy: 0.9704, Validation Accuracy: 0.9553, Loss: 0.0262
Epoch  38 Batch  240/1077 - Train Accuracy: 0.9875, Validation Accuracy: 0.9666, Loss: 0.0158
Epoch  38 Batch  260/1077 - Train Accuracy: 0.9777, Validation Accuracy: 0.9585, Loss: 0.0152
Epoch  38 Batch  280/1077 - Train Accuracy: 0.9539, Validation Accuracy: 0.9673, Loss: 0.0231
Epoch  38 Batch  300/1077 - Train Accuracy: 0.9753, Validation Accuracy: 0.9680, Loss: 0.0233
Epoch  38 Batch  320/1077 - Train Accuracy: 0.9539, Validation Accuracy: 0.9659, Loss: 0.0304
Epoch  38 Batch  340/1077 - Train Accuracy: 0.9749, Validation Accuracy: 0.9528, Loss: 0.0192
Epoch  38 Batch  360/1077 - Train Accuracy: 0.9660, Validation Accuracy: 0.9521, Loss: 0.0132
Epoch  38 Batch  380/1077 - Train Accuracy: 0.9914, Validation Accuracy: 0.9542, Loss: 0.0129
Epoch  38 Batch  400/1077 - Train Accuracy: 0.9496, Validation Accuracy: 0.9510, Loss: 0.0220
Epoch  38 Batch  420/1077 - Train Accuracy: 0.9938, Validati

Epoch  39 Batch  920/1077 - Train Accuracy: 0.9543, Validation Accuracy: 0.9719, Loss: 0.0192
Epoch  39 Batch  940/1077 - Train Accuracy: 0.9695, Validation Accuracy: 0.9631, Loss: 0.0194
Epoch  39 Batch  960/1077 - Train Accuracy: 0.9531, Validation Accuracy: 0.9574, Loss: 0.0178
Epoch  39 Batch  980/1077 - Train Accuracy: 0.9598, Validation Accuracy: 0.9627, Loss: 0.0205
Epoch  39 Batch 1000/1077 - Train Accuracy: 0.9621, Validation Accuracy: 0.9602, Loss: 0.0219
Epoch  39 Batch 1020/1077 - Train Accuracy: 0.9957, Validation Accuracy: 0.9599, Loss: 0.0113
Epoch  39 Batch 1040/1077 - Train Accuracy: 0.9708, Validation Accuracy: 0.9709, Loss: 0.0193
Epoch  39 Batch 1060/1077 - Train Accuracy: 0.9762, Validation Accuracy: 0.9716, Loss: 0.0140
Epoch  40 Batch   20/1077 - Train Accuracy: 0.9715, Validation Accuracy: 0.9730, Loss: 0.0146
Epoch  40 Batch   40/1077 - Train Accuracy: 0.9938, Validation Accuracy: 0.9670, Loss: 0.0161
Epoch  40 Batch   60/1077 - Train Accuracy: 0.9565, Validati

Epoch  41 Batch  560/1077 - Train Accuracy: 0.9590, Validation Accuracy: 0.9652, Loss: 0.0163
Epoch  41 Batch  580/1077 - Train Accuracy: 0.9617, Validation Accuracy: 0.9688, Loss: 0.0138
Epoch  41 Batch  600/1077 - Train Accuracy: 0.9740, Validation Accuracy: 0.9677, Loss: 0.0203
Epoch  41 Batch  620/1077 - Train Accuracy: 0.9715, Validation Accuracy: 0.9673, Loss: 0.0192
Epoch  41 Batch  640/1077 - Train Accuracy: 0.9777, Validation Accuracy: 0.9631, Loss: 0.0499
Epoch  41 Batch  660/1077 - Train Accuracy: 0.9785, Validation Accuracy: 0.9645, Loss: 0.0142
Epoch  41 Batch  680/1077 - Train Accuracy: 0.9565, Validation Accuracy: 0.9620, Loss: 0.0221
Epoch  41 Batch  700/1077 - Train Accuracy: 0.9840, Validation Accuracy: 0.9585, Loss: 0.0146
Epoch  41 Batch  720/1077 - Train Accuracy: 0.9494, Validation Accuracy: 0.9684, Loss: 0.0199
Epoch  41 Batch  740/1077 - Train Accuracy: 0.9680, Validation Accuracy: 0.9663, Loss: 0.0147
Epoch  41 Batch  760/1077 - Train Accuracy: 0.9809, Validati

Epoch  43 Batch  200/1077 - Train Accuracy: 0.9730, Validation Accuracy: 0.9577, Loss: 0.0154
Epoch  43 Batch  220/1077 - Train Accuracy: 0.9712, Validation Accuracy: 0.9531, Loss: 0.0197
Epoch  43 Batch  240/1077 - Train Accuracy: 0.9871, Validation Accuracy: 0.9698, Loss: 0.0132
Epoch  43 Batch  260/1077 - Train Accuracy: 0.9781, Validation Accuracy: 0.9712, Loss: 0.0132
Epoch  43 Batch  280/1077 - Train Accuracy: 0.9531, Validation Accuracy: 0.9627, Loss: 0.0202
Epoch  43 Batch  300/1077 - Train Accuracy: 0.9774, Validation Accuracy: 0.9723, Loss: 0.0155
Epoch  43 Batch  320/1077 - Train Accuracy: 0.9742, Validation Accuracy: 0.9677, Loss: 0.0260
Epoch  43 Batch  340/1077 - Train Accuracy: 0.9815, Validation Accuracy: 0.9670, Loss: 0.0179
Epoch  43 Batch  360/1077 - Train Accuracy: 0.9609, Validation Accuracy: 0.9638, Loss: 0.0146
Epoch  43 Batch  380/1077 - Train Accuracy: 0.9918, Validation Accuracy: 0.9599, Loss: 0.0117
Epoch  43 Batch  400/1077 - Train Accuracy: 0.9531, Validati

Epoch  44 Batch  900/1077 - Train Accuracy: 0.9746, Validation Accuracy: 0.9641, Loss: 0.0177
Epoch  44 Batch  920/1077 - Train Accuracy: 0.9746, Validation Accuracy: 0.9744, Loss: 0.0174
Epoch  44 Batch  940/1077 - Train Accuracy: 0.9707, Validation Accuracy: 0.9691, Loss: 0.0143
Epoch  44 Batch  960/1077 - Train Accuracy: 0.9777, Validation Accuracy: 0.9570, Loss: 0.0151
Epoch  44 Batch  980/1077 - Train Accuracy: 0.9652, Validation Accuracy: 0.9744, Loss: 0.0184
Epoch  44 Batch 1000/1077 - Train Accuracy: 0.9691, Validation Accuracy: 0.9616, Loss: 0.0163
Epoch  44 Batch 1020/1077 - Train Accuracy: 0.9949, Validation Accuracy: 0.9620, Loss: 0.0110
Epoch  44 Batch 1040/1077 - Train Accuracy: 0.9749, Validation Accuracy: 0.9712, Loss: 0.0169
Epoch  44 Batch 1060/1077 - Train Accuracy: 0.9766, Validation Accuracy: 0.9702, Loss: 0.0119
Epoch  45 Batch   20/1077 - Train Accuracy: 0.9719, Validation Accuracy: 0.9819, Loss: 0.0140
Epoch  45 Batch   40/1077 - Train Accuracy: 0.9832, Validati

Epoch  46 Batch  540/1077 - Train Accuracy: 0.9688, Validation Accuracy: 0.9677, Loss: 0.0133
Epoch  46 Batch  560/1077 - Train Accuracy: 0.9625, Validation Accuracy: 0.9695, Loss: 0.0178
Epoch  46 Batch  580/1077 - Train Accuracy: 0.9673, Validation Accuracy: 0.9677, Loss: 0.0125
Epoch  46 Batch  600/1077 - Train Accuracy: 0.9788, Validation Accuracy: 0.9659, Loss: 0.0173
Epoch  46 Batch  620/1077 - Train Accuracy: 0.9766, Validation Accuracy: 0.9790, Loss: 0.0163
Epoch  46 Batch  640/1077 - Train Accuracy: 0.9821, Validation Accuracy: 0.9766, Loss: 0.0153
Epoch  46 Batch  660/1077 - Train Accuracy: 0.9902, Validation Accuracy: 0.9705, Loss: 0.0092
Epoch  46 Batch  680/1077 - Train Accuracy: 0.9710, Validation Accuracy: 0.9641, Loss: 0.0162
Epoch  46 Batch  700/1077 - Train Accuracy: 0.9793, Validation Accuracy: 0.9602, Loss: 0.0132
Epoch  46 Batch  720/1077 - Train Accuracy: 0.9679, Validation Accuracy: 0.9677, Loss: 0.0152
Epoch  46 Batch  740/1077 - Train Accuracy: 0.9758, Validati

Epoch  48 Batch  180/1077 - Train Accuracy: 0.9738, Validation Accuracy: 0.9613, Loss: 0.0129
Epoch  48 Batch  200/1077 - Train Accuracy: 0.9770, Validation Accuracy: 0.9684, Loss: 0.0123
Epoch  48 Batch  220/1077 - Train Accuracy: 0.9786, Validation Accuracy: 0.9730, Loss: 0.0167
Epoch  48 Batch  240/1077 - Train Accuracy: 0.9867, Validation Accuracy: 0.9702, Loss: 0.0114
Epoch  48 Batch  260/1077 - Train Accuracy: 0.9740, Validation Accuracy: 0.9666, Loss: 0.0107
Epoch  48 Batch  280/1077 - Train Accuracy: 0.9563, Validation Accuracy: 0.9680, Loss: 0.0197
Epoch  48 Batch  300/1077 - Train Accuracy: 0.9782, Validation Accuracy: 0.9840, Loss: 0.0137
Epoch  48 Batch  320/1077 - Train Accuracy: 0.9676, Validation Accuracy: 0.9741, Loss: 0.0221
Epoch  48 Batch  340/1077 - Train Accuracy: 0.9811, Validation Accuracy: 0.9705, Loss: 0.0105
Epoch  48 Batch  360/1077 - Train Accuracy: 0.9734, Validation Accuracy: 0.9762, Loss: 0.0118
Epoch  48 Batch  380/1077 - Train Accuracy: 0.9898, Validati

Epoch  49 Batch  880/1077 - Train Accuracy: 0.9703, Validation Accuracy: 0.9428, Loss: 0.0191
Epoch  49 Batch  900/1077 - Train Accuracy: 0.9766, Validation Accuracy: 0.9663, Loss: 0.0169
Epoch  49 Batch  920/1077 - Train Accuracy: 0.9641, Validation Accuracy: 0.9698, Loss: 0.0185
Epoch  49 Batch  940/1077 - Train Accuracy: 0.9852, Validation Accuracy: 0.9691, Loss: 0.0122
Epoch  49 Batch  960/1077 - Train Accuracy: 0.9795, Validation Accuracy: 0.9666, Loss: 0.0155
Epoch  49 Batch  980/1077 - Train Accuracy: 0.9707, Validation Accuracy: 0.9744, Loss: 0.0170
Epoch  49 Batch 1000/1077 - Train Accuracy: 0.9814, Validation Accuracy: 0.9712, Loss: 0.0142
Epoch  49 Batch 1020/1077 - Train Accuracy: 0.9965, Validation Accuracy: 0.9730, Loss: 0.0076
Epoch  49 Batch 1040/1077 - Train Accuracy: 0.9848, Validation Accuracy: 0.9766, Loss: 0.0149
Epoch  49 Batch 1060/1077 - Train Accuracy: 0.9844, Validation Accuracy: 0.9769, Loss: 0.0101
Epoch  50 Batch   20/1077 - Train Accuracy: 0.9824, Validati

Epoch  51 Batch  520/1077 - Train Accuracy: 0.9885, Validation Accuracy: 0.9585, Loss: 0.0105
Epoch  51 Batch  540/1077 - Train Accuracy: 0.9754, Validation Accuracy: 0.9624, Loss: 0.0102
Epoch  51 Batch  560/1077 - Train Accuracy: 0.9625, Validation Accuracy: 0.9666, Loss: 0.0128
Epoch  51 Batch  580/1077 - Train Accuracy: 0.9781, Validation Accuracy: 0.9641, Loss: 0.0099
Epoch  51 Batch  600/1077 - Train Accuracy: 0.9773, Validation Accuracy: 0.9759, Loss: 0.0162
Epoch  51 Batch  620/1077 - Train Accuracy: 0.9766, Validation Accuracy: 0.9702, Loss: 0.0166
Epoch  51 Batch  640/1077 - Train Accuracy: 0.9769, Validation Accuracy: 0.9719, Loss: 0.0142
Epoch  51 Batch  660/1077 - Train Accuracy: 0.9871, Validation Accuracy: 0.9709, Loss: 0.0091
Epoch  51 Batch  680/1077 - Train Accuracy: 0.9680, Validation Accuracy: 0.9677, Loss: 0.0158
Epoch  51 Batch  700/1077 - Train Accuracy: 0.9805, Validation Accuracy: 0.9560, Loss: 0.0117
Epoch  51 Batch  720/1077 - Train Accuracy: 0.9725, Validati

Epoch  53 Batch  160/1077 - Train Accuracy: 0.9781, Validation Accuracy: 0.9730, Loss: 0.0122
Epoch  53 Batch  180/1077 - Train Accuracy: 0.9855, Validation Accuracy: 0.9705, Loss: 0.0113
Epoch  53 Batch  200/1077 - Train Accuracy: 0.9797, Validation Accuracy: 0.9808, Loss: 0.0103
Epoch  53 Batch  220/1077 - Train Accuracy: 0.9782, Validation Accuracy: 0.9471, Loss: 0.0152
Epoch  53 Batch  240/1077 - Train Accuracy: 0.9867, Validation Accuracy: 0.9759, Loss: 0.0104
Epoch  53 Batch  260/1077 - Train Accuracy: 0.9821, Validation Accuracy: 0.9712, Loss: 0.0105
Epoch  53 Batch  280/1077 - Train Accuracy: 0.9645, Validation Accuracy: 0.9727, Loss: 0.0166
Epoch  53 Batch  300/1077 - Train Accuracy: 0.9766, Validation Accuracy: 0.9822, Loss: 0.0108
Epoch  53 Batch  320/1077 - Train Accuracy: 0.9703, Validation Accuracy: 0.9762, Loss: 0.0198
Epoch  53 Batch  340/1077 - Train Accuracy: 0.9840, Validation Accuracy: 0.9751, Loss: 0.0110
Epoch  53 Batch  360/1077 - Train Accuracy: 0.9762, Validati

Epoch  54 Batch  860/1077 - Train Accuracy: 0.9859, Validation Accuracy: 0.9805, Loss: 0.0145
Epoch  54 Batch  880/1077 - Train Accuracy: 0.9891, Validation Accuracy: 0.9428, Loss: 0.0186
Epoch  54 Batch  900/1077 - Train Accuracy: 0.9777, Validation Accuracy: 0.9609, Loss: 0.0153
Epoch  54 Batch  920/1077 - Train Accuracy: 0.9836, Validation Accuracy: 0.9688, Loss: 0.0116
Epoch  54 Batch  940/1077 - Train Accuracy: 0.9848, Validation Accuracy: 0.9730, Loss: 0.0109
Epoch  54 Batch  960/1077 - Train Accuracy: 0.9676, Validation Accuracy: 0.9702, Loss: 0.0133
Epoch  54 Batch  980/1077 - Train Accuracy: 0.9758, Validation Accuracy: 0.9719, Loss: 0.0140
Epoch  54 Batch 1000/1077 - Train Accuracy: 0.9781, Validation Accuracy: 0.9744, Loss: 0.0143
Epoch  54 Batch 1020/1077 - Train Accuracy: 0.9965, Validation Accuracy: 0.9744, Loss: 0.0083
Epoch  54 Batch 1040/1077 - Train Accuracy: 0.9889, Validation Accuracy: 0.9741, Loss: 0.0193
Epoch  54 Batch 1060/1077 - Train Accuracy: 0.9871, Validati

Epoch  56 Batch  500/1077 - Train Accuracy: 0.9750, Validation Accuracy: 0.9684, Loss: 0.0081
Epoch  56 Batch  520/1077 - Train Accuracy: 0.9892, Validation Accuracy: 0.9663, Loss: 0.0087
Epoch  56 Batch  540/1077 - Train Accuracy: 0.9723, Validation Accuracy: 0.9695, Loss: 0.0107
Epoch  56 Batch  560/1077 - Train Accuracy: 0.9621, Validation Accuracy: 0.9762, Loss: 0.0147
Epoch  56 Batch  580/1077 - Train Accuracy: 0.9740, Validation Accuracy: 0.9748, Loss: 0.0097
Epoch  56 Batch  600/1077 - Train Accuracy: 0.9859, Validation Accuracy: 0.9719, Loss: 0.0173
Epoch  56 Batch  620/1077 - Train Accuracy: 0.9766, Validation Accuracy: 0.9805, Loss: 0.0146
Epoch  56 Batch  640/1077 - Train Accuracy: 0.9732, Validation Accuracy: 0.9734, Loss: 0.0126
Epoch  56 Batch  660/1077 - Train Accuracy: 0.9855, Validation Accuracy: 0.9741, Loss: 0.0089
Epoch  56 Batch  680/1077 - Train Accuracy: 0.9710, Validation Accuracy: 0.9638, Loss: 0.0139
Epoch  56 Batch  700/1077 - Train Accuracy: 0.9895, Validati

Epoch  58 Batch  140/1077 - Train Accuracy: 0.9893, Validation Accuracy: 0.9645, Loss: 0.0096
Epoch  58 Batch  160/1077 - Train Accuracy: 0.9820, Validation Accuracy: 0.9677, Loss: 0.0096
Epoch  58 Batch  180/1077 - Train Accuracy: 0.9797, Validation Accuracy: 0.9702, Loss: 0.0103
Epoch  58 Batch  200/1077 - Train Accuracy: 0.9859, Validation Accuracy: 0.9723, Loss: 0.0110
Epoch  58 Batch  220/1077 - Train Accuracy: 0.9774, Validation Accuracy: 0.9474, Loss: 0.0140
Epoch  58 Batch  240/1077 - Train Accuracy: 0.9895, Validation Accuracy: 0.9812, Loss: 0.0104
Epoch  58 Batch  260/1077 - Train Accuracy: 0.9781, Validation Accuracy: 0.9744, Loss: 0.0087
Epoch  58 Batch  280/1077 - Train Accuracy: 0.9602, Validation Accuracy: 0.9705, Loss: 0.0135
Epoch  58 Batch  300/1077 - Train Accuracy: 0.9852, Validation Accuracy: 0.9695, Loss: 0.0101
Epoch  58 Batch  320/1077 - Train Accuracy: 0.9770, Validation Accuracy: 0.9762, Loss: 0.0165
Epoch  58 Batch  340/1077 - Train Accuracy: 0.9856, Validati

Epoch  59 Batch  840/1077 - Train Accuracy: 0.9926, Validation Accuracy: 0.9744, Loss: 0.0108
Epoch  59 Batch  860/1077 - Train Accuracy: 0.9866, Validation Accuracy: 0.9677, Loss: 0.0132
Epoch  59 Batch  880/1077 - Train Accuracy: 0.9918, Validation Accuracy: 0.9613, Loss: 0.0132
Epoch  59 Batch  900/1077 - Train Accuracy: 0.9836, Validation Accuracy: 0.9680, Loss: 0.0157
Epoch  59 Batch  920/1077 - Train Accuracy: 0.9805, Validation Accuracy: 0.9712, Loss: 0.0113
Epoch  59 Batch  940/1077 - Train Accuracy: 0.9891, Validation Accuracy: 0.9737, Loss: 0.0092
Epoch  59 Batch  960/1077 - Train Accuracy: 0.9699, Validation Accuracy: 0.9734, Loss: 0.0117
Epoch  59 Batch  980/1077 - Train Accuracy: 0.9758, Validation Accuracy: 0.9705, Loss: 0.0136
Epoch  59 Batch 1000/1077 - Train Accuracy: 0.9788, Validation Accuracy: 0.9723, Loss: 0.0112
Epoch  59 Batch 1020/1077 - Train Accuracy: 0.9961, Validation Accuracy: 0.9645, Loss: 0.0054
Epoch  59 Batch 1040/1077 - Train Accuracy: 0.9844, Validati

### Save Parameters
Save the `batch_size` and `save_path` parameters for inference.

In [37]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
# Save parameters for checkpoint
helper.save_params(save_path)

# Checkpoint

In [38]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import tensorflow as tf
import numpy as np
import helper
import problem_unittests as tests

_, (source_vocab_to_int, target_vocab_to_int), (source_int_to_vocab, target_int_to_vocab) = helper.load_preprocess()
load_path = helper.load_params()

## Sentence to Sequence
To feed a sentence into the model for translation, you first need to preprocess it.  Implement the function `sentence_to_seq()` to preprocess new sentences.

- Convert the sentence to lowercase
- Convert words into ids using `vocab_to_int`
 - Convert words not in the vocabulary, to the `<UNK>` word id.

In [50]:
def sentence_to_seq(sentence, vocab_to_int):
    """
    Convert a sentence to a sequence of ids
    :param sentence: String
    :param vocab_to_int: Dictionary to go from the words to an id
    :return: List of word ids
    """
    return [vocab_to_int.get(word, vocab_to_int['<UNK>']) for word in sentence.split(' ')]

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_sentence_to_seq(sentence_to_seq)

Tests Passed


## Translate
This will translate `translate_sentence` from English to French.

In [54]:
translate_sentence = 'he saw a old yellow truck .'


"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
translate_sentence = sentence_to_seq(translate_sentence, source_vocab_to_int)

loaded_graph = tf.Graph()
with tf.Session(graph=loaded_graph) as sess:
    # Load saved model
    loader = tf.train.import_meta_graph(load_path + '.meta')
    loader.restore(sess, load_path)

    input_data = loaded_graph.get_tensor_by_name('input:0')
    logits = loaded_graph.get_tensor_by_name('predictions:0')
    target_sequence_length = loaded_graph.get_tensor_by_name('target_sequence_length:0')
    source_sequence_length = loaded_graph.get_tensor_by_name('source_sequence_length:0')
    keep_prob = loaded_graph.get_tensor_by_name('keep_prob:0')

    translate_logits = sess.run(logits, {input_data: [translate_sentence]*batch_size,
                                         target_sequence_length: [len(translate_sentence)*2]*batch_size,
                                         source_sequence_length: [len(translate_sentence)]*batch_size,
                                         keep_prob: 1.0})[0]

print('Input')
print('  Word Ids:      {}'.format([i for i in translate_sentence]))
print('  English Words: {}'.format([source_int_to_vocab[i] for i in translate_sentence]))

print('\nPrediction')
print('  Word Ids:      {}'.format([i for i in translate_logits]))
print('  French Words: {}'.format(" ".join([target_int_to_vocab[i] for i in translate_logits])))


INFO:tensorflow:Restoring parameters from checkpoints/dev
Input
  Word Ids:      [101, 174, 166, 124, 131, 139, 199]
  English Words: ['he', 'saw', 'a', 'old', 'yellow', 'truck', '.']

Prediction
  Word Ids:      [331, 41, 300, 109, 252, 252, 84, 1]
  French Words: a vu un vieux jaune jaune . <EOS>


## Imperfect Translation
You might notice that some sentences translate better than others.  Since the dataset you're using only has a vocabulary of 227 English words of the thousands that you use, you're only going to see good results using these words.  For this project, you don't need a perfect translation. However, if you want to create a better translation model, you'll need better data.

You can train on the [WMT10 French-English corpus](http://www.statmt.org/wmt10/training-giga-fren.tar).  This dataset has more vocabulary and richer in topics discussed.  However, this will take you days to train, so make sure you've a GPU and the neural network is performing well on dataset we provided.  Just make sure you play with the WMT10 corpus after you've submitted this project.
## Submitting This Project
When submitting this project, make sure to run all the cells before saving the notebook. Save the notebook file as "dlnd_language_translation.ipynb" and save it as a HTML file under "File" -> "Download as". Include the "helper.py" and "problem_unittests.py" files in your submission.